Address problems with binomial approximation

Consider this simple accumulator model that samples from a binomial each time step and returns the time step when the sum exceeds a threshold. 

```
var n = 100

var accumulator = function(state, t) {
  var newState = binomial(.05, n);
  if(state + newState >= n) {
    return t;
  } else {
    return accumulator(state + newState, t+1);
  }
};

var model = function(){
  return accumulator(0, 0);
};

print(EnumerateBreadthFirst(model,  1))
```

Note that without doing breadth-first search, this will not terminate and will hit a memory error. However, using breadth-first search throws an error: ```webppl.min.js:2 Uncaught AssertionError: false == true```. Maybe this will be clearer once we have the stack traces, but the weird part is that you only get the error when `n` is set to `101` or higher. When it's `100` or less, it works as expected. I'm guessing this has something to do with the binomial approximation?

> And it's got bugs -- (e.g., score(Binomial({n: 21, p: 0.5}), 21) returns -Infinity). 

I'm not sure I would call that a bug, the true log prob is something like -300, so -Infinity seems like a pretty good approximation! :)

> using breadth-first search throws an error: webppl.min.js:2 Uncaught AssertionError: false == true.

I was intrigued by what exactly was happening here. It turns out that the underlying problem is that the single execution that finishes has probability zero (because of the approximation @longouyang pointed out) so it's not possible to return a normalized marginal. We [made the error message here more useful](https://github.com/probmods/webppl/commit/1cd41fdb0a7106c1b1e198ae07e1fc7083291920) a while ago, looks like that hasn't made it to webppl.org yet.

> I'm not sure if the approximation buys us anything

--- a/src/dists.ad.js
+++ b/src/dists.ad.js
@@ -574,6 +574,8 @@ function binomialG(x) {
   return (1 - (x * x) + (2 * x * Math.log(x))) / (d * d);
 }
 
+// see lemma 6.1 from Ahrens & Dieter's
+// Computer Methods for Sampling from Gamma, Beta, Poisson and Binomial Distributions
 function binomialSample(p, n) {
   var k = 0;
   var N = 10;

--- a/src/dists.ad.js
+++ b/src/dists.ad.js
@@ -614,32 +614,9 @@ var Binomial = makeDistributionType({
     'use ad';
     var p = this.params.p;
     var n = this.params.n;
-    if (n > 20 && n * p > 5 && n * (1 - p) > 5) {
-      // large n, reasonable p approximation
-      var s = val;
-      var inv2 = 1 / 2;
-      var inv3 = 1 / 3;
-      var inv6 = 1 / 6;
-      if (s >= n) {
-        return -Infinity;
-      }
-      var q = 1 - p;
-      var S = s + inv2;
-      var T = n - s - inv2;
-      var d1 = s + inv6 - (n + inv3) * p;
-      var d2 = q / (s + inv2) - p / (T + inv2) + (q - inv2) / (n + 1);
-      d2 = d1 + 0.02 * d2;
-      var num = 1 + q * binomialG(S / (n * p)) + p * binomialG(T / (n * q));
-      var den = (n + inv6) * p * q;
-      var z = num / den;
-      var invsd = Math.sqrt(z);
-      z = d2 * invsd;
-      return gaussianScore(0, 1, z) + Math.log(invsd);
-    } else {
-      // exact formula
-      return (lnfact(n) - lnfact(n - val) - lnfact(val) +
-          val * Math.log(p) + (n - val) * Math.log(1 - p));
-    }
+    // exact formula
+    return (lnfact(n) - lnfact(n - val) - lnfact(val) +
+            val * Math.log(p) + (n - val) * Math.log(1 - p));
   },
   support: function() {
     return _.range(this.params.n).concat(this.params.n);

